multipaz-models/src/commonMain/kotlin/org/multipaz/models/provisioning/ProvisioningModel.kt

```diff
@@ -0,0 +1,353 @@
+package org.multipaz.models.provisioning
+
+import io.ktor.client.HttpClient
+import kotlinx.coroutines.CancellationException
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Deferred
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.async
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.io.bytestring.encodeToByteString
+import kotlinx.serialization.json.buildJsonObject
+import kotlinx.serialization.json.put
+import org.multipaz.credential.Credential
+import org.multipaz.credential.SecureAreaBoundCredential
+import org.multipaz.document.AbstractDocumentMetadata
+import org.multipaz.document.Document
+import org.multipaz.document.DocumentStore
+import org.multipaz.mdoc.credential.MdocCredential
+import org.multipaz.prompt.PromptModel
+import org.multipaz.provision.AuthorizationChallenge
+import org.multipaz.provision.AuthorizationResponse
+import org.multipaz.provision.CredentialFormat
+import org.multipaz.provision.Display
+import org.multipaz.provision.KeyBindingInfo
+import org.multipaz.provision.KeyBindingType
+import org.multipaz.provision.ProvisioningClient
+import org.multipaz.provision.openid4vci.OpenID4VCI
+import org.multipaz.provision.openid4vci.OpenID4VCIBackend
+import org.multipaz.provision.openid4vci.OpenID4VCIClientPreferences
+import org.multipaz.rpc.backend.BackendEnvironment
+import org.multipaz.sdjwt.credential.KeyBoundSdJwtVcCredential
+import org.multipaz.sdjwt.credential.KeylessSdJwtVcCredential
+import org.multipaz.securearea.CreateKeySettings
+import org.multipaz.securearea.SecureArea
+import org.multipaz.securearea.SecureAreaProvider
+import org.multipaz.storage.Storage
+import org.multipaz.util.Logger
+import org.multipaz.util.toBase64Url
+import kotlin.coroutines.CoroutineContext
+import kotlin.math.min
+import kotlin.reflect.KClass
+import kotlin.reflect.safeCast
+import kotlin.time.Clock
+import kotlin.time.Duration.Companion.days
+
+/**
+ * This model supports UX/UI flow for provisioning of credentials.
+ *
+ * Only a single provisioning session per model object can be active at any time.
+ */
+class ProvisioningModel(
+    private val documentStore: DocumentStore,
+    private val secureArea: SecureArea,
+    private val httpClient: HttpClient,
```

Comment:
Should this take a `httpClientEngineFactory: HttpClientEngineFactory<*>` instead?